presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,250 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.server;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.prestosql.execution.SqlQueryExecution;
+import io.prestosql.execution.StageInfo;
+import io.prestosql.execution.StageState;
+import io.prestosql.execution.TaskInfo;
+import io.prestosql.execution.TaskManagerConfig;
+import io.prestosql.spi.QueryId;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.predicate.Domain;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.sql.DynamicFilters;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.SymbolReference;
+
+import javax.annotation.PreDestroy;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.Immutable;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.MoreFutures.addSuccessCallback;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+
+@ThreadSafe
+public class DynamicFilterService
+{
+    @GuardedBy("this")
+    private final Map<SourceDescriptor, SettableFuture<Domain>> dynamicFilterSummaries = new HashMap<>();
```

Comment:
I've made dynamicFilterSummaries `ConcurrentHashMap` and removed `synchronized` from `getSummaryFuture`.
Using Future here helps me to setup a callback in createDynamicFilterSupplier which can asynchronously `translateSummaryToTupleDomain` once and update the supplier. Supplier.get remains a cheap operation (get on AtomicReference) as it doesn't access the map again after the initial setup in createDynamicFilterSupplier.
If I store `Domain` instead of future, then each Supplier.get operation would need to perform lookups on dynamicFilterSummaries map. We'll also need additional logic here to avoid repeating the `translateSummaryToTupleDomain` and `intersect` calls, current code is able to avoid repeating work by using callbacks.

Reply:
> If I store Domain instead of future, then each Supplier.get operation would need to perform lookups on dynamicFilterSummaries map

That should be fine since `ConcurrentHashMap` has `O(1)` get access time. This supplier is called once per partition or split so it's not that critical. However, the code here would get simpler.

Reply:
I've now simplified the code to use `Optional<Domain>` instead of `SettableFuture<Domain>`. I'm using the `Optional` to figure out whether dynamic filters have already been collected for a filterId and avoiding redoing the work for them in `collectDynamicFilters`.